Only respond to modifier-only keybindings on release

spl237 opened the pull request:
This is a possible fix for https://github.com/labwc/labwc/issues/1756

I have modified the keybinding code so that a flag is set in the keybinding data structure if the keybinding only involves modifier keys.

If this flag is set for a particular keybinding, then that keybinding is detected on key press and stored, and executed on key release if no other (non-modifier) keybinding has executed in the meantime.

This fixes my immediate problem, that of wanting Win+cursor to activate snap behaviour, and Win on its own to launch the application menu, but I have no idea if this is the best way to do this. It seems to work for me, though!

Consolatis commented on src/input/keyboard.c:
```diff
@@ -407,7 +409,14 @@ handle_compositor_keybindings(struct keyboard *keyboard,
-		return handle_key_release(server, event->keycode);
+		if (cur_keybind && cur_keybind->mod_only) {
+			key_state_store_pressed_key_as_bound(event->keycode);
+			actions_run(NULL, server, &cur_keybind->actions, 0);
```
This would allow modifier only keybinds to be execute when the session is locked.

spl237 replied:
OK, that's not an area I understand - can you please explain a bit more, and I will see if I can fix it?

Consolatis replied:
The check for being locked (or keyboard inhibited) comes a bit further below as that was only be relevant for the press events before. It might be enough to just guard the `actions_run()` and `key_state_store_pressed_key_as_bound()` with a `if (!>server->session_lock && !seat->active_client_while_inhibited)`.

spl237 replied:
Ah yes - I see what you mean, thanks. I'll change accordingly.

Consolatis commented:
I am also not sure the key tracking is correct here, if we eat the release event but let the press event through to the focused client that will likely result in the client repeating the key on a timer.

spl237 commented:
Agree this shouldn't be in 0.7.2 - it's a quick hack I put together this morning, and needs a lot of testing for corner cases etc, but if the general approach is ok, I'm happy to look at improvements.

spl237 commented:
> I am also not sure the key state tracking is correct here, if we eat the release event but let the press event through to the focused client that will likely result in the client repeating the key on a timer.

I'm not sure I completely understand the key state tracking, but in an attempt to deal with this, I have modified the code so that the stored key state is only updated on press and not on release - does this look like a better approach?

Interestingly, it has also meant that a modifier-only keybinding is only executed if the key is released relatively quickly, not if it is held down for more than a second or so, which I think is actually better.

Consolatis commented:
> There is also another, older issue for modifier-only keybinds #351 ^ Just remembered that one and thought I'd mention it, so it gets closed if this PR fixes the issue.

For some reason I can't link that issue to the PR (it doesn't show up in the link issue popup). Well, guess we'll have to manually close it then once this is merged.

ahesford commented:
Linked. The issue just had to be pinged to convince GitHub that it was still active.

johanmalm commented:
Sorry it's taken a while to get to this one. Just working down the list.

What's the best way of testing it? I tried binding to `W` but can't get it to work.

```
    <keybind key="W">
      <action name="Execute" command="rofi -show drun"/>
    </keybind>
```

spl237 commented:
Bind to Super_L:

`    <keybind key="Super_L">
      <action name="Execute">
        <command>wfpanelctl smenu menu</command>
      </action>
    </keybind>`

johanmalm commented:
> Bind to Super_L:
> 
> ` <keybind key="Super_L"> <action name="Execute"> <command>wfpanelctl smenu menu</command> </action> </keybind>`
> 
> I think W is the key when used as a modifier, and Super_L is the same key as a plain keystroke.

Genius. Will have a play :smile:

johanmalm commented on src/input/keyboard.c:
```diff
@@ -407,7 +409,18 @@ handle_compositor_keybindings(struct keyboard *keyboard,
+			if (seat->active_client_while_inhibited
+					|| seat->server->session_lock) {
+				cur_keybind = NULL;
+				return false;
```
I think should be `true` to that the event is not passed onto the client.

spl237 replied:
Done

johanmalm commented on src/input/keyboard.c:
```diff
@@ -407,7 +409,18 @@ handle_compositor_keybindings(struct keyboard *keyboard,
+		if (cur_keybind && cur_keybind->mod_only) {
+			if (seat->active_client_while_inhibited
+					|| seat->server->session_lock) {
+				cur_keybind = NULL;
```
We store key as bound (https://github.com/labwc/labwc/blob/1cd9940307c4b54503cefcd643e92314ade88c45/src/input/keyboard.c#L454) so need to call `key_state_bound_key_remove()` here.

spl237 replied:
Done

johanmalm commented:
I've tested and it works well.

I'm pretty close to merging, but feel we ought to consider a few points.

1. `s/TRUE/true/` and `s/FALSE/false/`. I think the capitalized versions are some glib thing from the pre `stdbool.h` era.

2. I suggest changing `mod_only` to `on_release`. I think it's easier for the reader to understand what's going on.

3. Is is better to be explicit about when we handle the `release` event. For example: `<keybind key="Super_L" event="onRelease">` instead of calling `keyboard_is_modifier_key()`. I think it's pretty unlikely that anyone will want to bind to the press-event of a modifier, but you can't beat spelling it out. I can have a go at doing the parsing if you want, but let's have the discussion first.

4. Consider if we should allow multiple modifiers. It looks like it currently does (although not tested). I think the main use-case is `Super_L` so we could just set a rule to only allow on-release binds with `keysyms_len==1`. It's likely to make the key-state logic simpler too.

5. I've followed the key-state logic through and made a couple of suggestions. When merge-ready, I can follow it through with `LABWC_DEBUG_KEY_STATE=1` to double-check. It's an area of the code where it's particularly easy to make thought-mistakes.

spl237 commented:
OK, thanks for that - will have a look at those changes tomorrow.

Consolatis commented on src/input/keyboard.c:
```diff
@@ -38,6 +38,8 @@ struct keyinfo {
 
 static bool should_cancel_cycling_on_next_key_release;
 
+struct keybind *cur_keybind;
```
This should be static.

I am also not sure about `Reconfigure` happening between press and release, this seems to miss some reconfigure hook to reset `cur_keybind` in that case.

spl237 replied:
Done

spl237 commented:
> I've tested and it works well.
> 
> I'm pretty close to merging, but feel we ought to consider a few points.
> 
> 1. `s/TRUE/true/` and `s/FALSE/false/`. I think the capitalized versions are some glib thing from the pre `stdbool.h` era.

Fixed.

> 2. I suggest changing `mod_only` to `on_release`. I think it's easier for the reader to understand what's going on.

Fixed.

> 3. Is is better to be explicit about when we handle the `release` event. For example: `<keybind key="Super_L" event="onRelease">` instead of calling `keyboard_is_modifier_key()`. I think it's pretty unlikely that anyone will want to bind to the press-event of a modifier, but you can't beat spelling it out. I can have a go at doing the parsing if you want, but let's have the discussion first.

That sounds sensible to me.

> 4. Consider if we should allow multiple modifiers. It looks like it currently does (although not tested). I think the main use-case is `Super_L` so we could just set a rule to only allow on-release binds with `keysyms_len==1`. It's likely to make the key-state logic simpler too.

I can't think of any situations where you would want to bind to multiple modifiers for a single action, but I've got nothing against the idea if others think it would be useful.

> 5. I've followed the key-state logic through and made a couple of suggestions. When merge-ready, I can follow it through with `LABWC_DEBUG_KEY_STATE=1` to double-check. It's an area of the code where it's particularly easy to make thought-mistakes.

I think those should be fixed, but do please check.

Many thanks!

Consolatis commented:
Merged via
- #1888
